pkg/network/proxy/hybridproxier.go

```diff
@@ -0,0 +1,451 @@
+package proxy
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/klog/v2"
+
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1beta1 "k8s.io/api/discovery/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/kubernetes/pkg/proxy"
+	"k8s.io/kubernetes/pkg/util/async"
+
+	unidlingapi "github.com/openshift/api/unidling/v1alpha1"
+)
+
+// HybridizableProxy is an extra interface we layer on top of Provider
+type HybridizableProxy interface {
+	proxy.Provider
+
+	SyncProxyRules()
+	SetSyncRunner(b *async.BoundedFrequencyRunner)
+}
+
+// hybridProxierService is our cached state for a given Service/Endpoints.
+//
+// A running Service can be in one of three states:
+//
+//   - Not Idled (known to the mainProxy but not the unidlingProxy). A Not Idled Service
+//     becomes Idled when its Service gets annotated and its Endpoints is empty. (Both
+//     conditions must be true.)
+//
+//   - Idled (known to the unidlingProxy but not the mainProxy). An Idled Service becomes
+//     Unidling when either its Service annotation is removed or its Endpoints become
+//     non-empty.
+//
+//   - Unidling (the Service and Endpoints are known to the mainProxy, and the Endpoints
+//     are known to the unidling proxy). While a Service is Unidling, Endpoints events are
+//     sent to both proxies (so the unidling proxy socket can redirect its connection to
+//     the correct place). An Unidling Service becomes Not Idled if its Endpoints are
+//     deleted, or else the next time it receives an Endpoints event more than 1 minute
+//     after becoming Unidling. (Alternatively it could also become Idled again.)
```

Comment:
This timeline is different from the previous behavior - why did you do this? Wouldn't a service transition from Unidling to Not Idled iff it has non-empty endpoints? I'm assuming it's because a service can't possible unidle for more than a minute, but does that actually help?

Also, why would deleting Endpoints make a service Not Idled? Just because it means the service is going away?

Reply:
The old behavior was that we always sent all Endpoints events to the unidling proxy. The new behavior is that we're trying to not do that, because we have to synthesize fake Endpoints objects from EndpointSlices, and we don't want to have to do that for every EndpointSlice and double our memory usage (especially since almost all of the undling Endpoints updates are pointless; the unidling proxy has no need to ever know the Endpoints of Services that never get idled).

So the goal is "figure out if the unidling proxy actually cares about the endpoints for this service or not, and if not, then don't bother synthesizing an Endpoints from the EndpointSlice if it doesn't actually need to know".

When a Service has just un-idled and the UnidlerSocket still has pending connections to proxy, we need to send it Endpoints updates so it can know where to proxy the connections to. Once it has connected all of its pending connections, then it doesn't need to know about any further Endpoints events. It's likely this will happen within a few seconds of it receiving the first non-empty Endpoints update.

> Wouldn't a service transition from Unidling to Not Idled iff it has non-empty endpoints?

If the Service is backed by a Deployment/ReplicaSet/etc, then when it un-idles, we will first get an Endpoints update saying it has 1 endpoint, quickly followed by another Endpoints update saying it has 2 endpoints, etc. We _could_ decide we were only going to send the unidling proxy the first update (so that all connections captured by the unidling proxy would get proxied to the first endpoint to come up) but it seemed better to send it the others too?

Reply:
> Also, why would deleting Endpoints make a service Not Idled? Just because it means the service is going away?

yes